torch/utils/_pytree.py

```diff
@@ -274,7 +274,19 @@ def tree_flatten(pytree: PyTree) -> Tuple[List[Any], TreeSpec]:
 
 def tree_leaves(pytree: PyTree) -> List[Any]:
     """Get a list of leaves of a pytree."""
-    return tree_flatten(pytree)[0]
+    if _is_leaf(pytree):
+        return [pytree]
+
+    node_type = _get_node_type(pytree)
+    flatten_fn = SUPPORTED_NODES[node_type].flatten_fn
+    child_pytrees, context = flatten_fn(pytree)
+
+    # Recursively flatten the children
+    result : List[Any] = []
+    for child in child_pytrees:
+        result += tree_leaves(child)
+
+    return result
```

Comment:
I think we could have an even further optimized implementation using generators. I haven't profiled this yet. The implementation in this PR will create intermediate `list` objects in the recursive `tree_leaves` call.

Reply:
Yeah you don't even need a generator, we can pass the result as an argument to a helper and append directly into the list.
```python
def tree_leaves_helper(pytree, result):
    if _isleaf(pytree):
        result.append(pytree)
    # [snip]
    for child in child_pytrees:
        tree_leaves_helper(child, result)
```

This way there are no intermediates at all. In my small benchmark it was only a 5% uplift, and I didn't want to diverge too far from the `tree_flatten` implementation so they would be easier to keep in sync. Could maybe change both though in a further PR.